node-api/backend/validator.go

```diff
@@ -21,80 +21,180 @@
 package backend
 
 import (
+	"bytes"
+	"slices"
+	"strconv"
+
+	"github.com/berachain/beacon-kit/consensus-types/types"
+	"github.com/berachain/beacon-kit/errors"
 	"github.com/berachain/beacon-kit/node-api/backend/utils"
 	beacontypes "github.com/berachain/beacon-kit/node-api/handlers/beacon/types"
+	"github.com/berachain/beacon-kit/primitives/crypto"
 	"github.com/berachain/beacon-kit/primitives/math"
+	statedb "github.com/berachain/beacon-kit/state-transition/core/state"
 )
 
+// FilteredValidators will grab all of the validators from the state at the
+// given slot. It will then filter them by the provided ids and statuses.
+func (b Backend) FilteredValidators(
+	slot math.Slot, ids []string, statuses []string,
+) ([]*beacontypes.ValidatorData, error) {
+	st, _, err := b.stateFromSlot(slot)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get state from slot %d", slot)
+	}
+
+	validators, err := st.GetValidators()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get validators")
+	}
+
+	validatorData := make([]*beacontypes.ValidatorData, 0, len(validators))
+	for _, validator := range validators {
+		data, errInProcess := b.processValidator(st, validator, slot, ids, statuses)
+		if errInProcess != nil {
+			return nil, errors.Wrapf(errInProcess, "failed to process validator")
+		}
+		if data != nil {
+			validatorData = append(validatorData, data)
+		}
+	}
+	return validatorData, nil
+}
+
+func (b Backend) processValidator(
+	st *statedb.StateDB,
+	validator *types.Validator,
+	slot math.Slot,
+	ids []string,
+	statuses []string,
+) (*beacontypes.ValidatorData, error) {
+	index, err := st.ValidatorIndexByPubkey(validator.GetPubkey())
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get validator index by pubkey")
+	}
+
+	if !b.matchesIDFilter(validator, index, ids) {
+		//nolint:nilnil // no data to return
+		return nil, nil
+	}
+
+	status, err := validator.Status(b.cs.SlotToEpoch(slot))
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get validator status")
+	}
+
+	if !b.matchesStatusFilter(status, statuses) {
+		//nolint:nilnil // no data to return
+		return nil, nil
+	}
+
+	return b.buildValidatorData(st, validator, index, status)
+}
+
+func (b Backend) matchesIDFilter(validator *types.Validator, index math.U64, ids []string) bool {
+	if len(ids) == 0 {
+		return true
+	}
+
+	for _, id := range ids {
+		if b.matchesPubkey(validator, id) || b.matchesIndex(index, id) {
+			return true
+		}
+	}
+	return false
+}
+
+func (b Backend) matchesPubkey(validator *types.Validator, id string) bool {
+	var pubkey crypto.BLSPubkey
+	if err := pubkey.UnmarshalText([]byte(id)); err != nil {
```

Comment:
Also here I think we want to turn strings into PubKeys only once before passing them to `processValidator`. Otherwise we risk to Unmarshal over and over the same pubkeys

Reply:
addressed, parsed pubKey as crypto.BLSPubkey and doing it before processValidator.